src/circuit/note_commit.rs

```diff
@@ -540,7 +689,149 @@ impl NoteCommitConfig {
             Constraints::with_selector(q_y_canon, decomposition_checks.chain(canonicity_checks))
         });
 
-        config
+        Self { q_y_canon }
+    }
+}
+
+#[allow(non_snake_case)]
+#[derive(Clone, Debug)]
+pub struct NoteCommitConfig {
+    b: DecomposeB,
+    d: DecomposeD,
+    e: DecomposeE,
+    g: DecomposeG,
+    h: DecomposeH,
+    g_d: GdCanonicity,
+    pk_d: PkdCanonicity,
+    value: ValueCanonicity,
+    rho: RhoCanonicity,
+    psi: PsiCanonicity,
+    y_canon: YCanonicity,
+    advices: [Column<Advice>; 10],
+    sinsemilla_config:
+        SinsemillaConfig<OrchardHashDomains, OrchardCommitDomains, OrchardFixedBases>,
+}
+
+impl NoteCommitConfig {
+    #[allow(non_snake_case)]
+    #[allow(clippy::many_single_char_names)]
+    pub(in crate::circuit) fn configure(
+        meta: &mut ConstraintSystem<pallas::Base>,
+        advices: [Column<Advice>; 10],
+        sinsemilla_config: SinsemillaConfig<
+            OrchardHashDomains,
+            OrchardCommitDomains,
+            OrchardFixedBases,
+        >,
+    ) -> Self {
+        // Useful constants
+        let two = pallas::Base::from(2);
+        let two_pow_2 = pallas::Base::from(1 << 2);
+        let two_pow_4 = two_pow_2.square();
+        let two_pow_5 = two_pow_4 * two;
+        let two_pow_6 = two_pow_5 * two;
+        let two_pow_8 = two_pow_4.square();
+        let two_pow_9 = two_pow_8 * two;
+        let two_pow_10 = two_pow_9 * two;
+        let two_pow_58 = pallas::Base::from(1 << 58);
+        let two_pow_130 = Expression::Constant(pallas::Base::from_u128(1 << 65).square());
+        let two_pow_140 = Expression::Constant(pallas::Base::from_u128(1 << 70).square());
+        let two_pow_249 = pallas::Base::from_u128(1 << 124).square() * two;
+        let two_pow_250 = two_pow_249 * two;
+        let two_pow_254 = pallas::Base::from_u128(1 << 127).square();
```

Comment:
It seems a bit unfortunate that there's a mix of `pallas::Base` and `Expression` here. Is there overhead to `Expression::Constant`? Why not uniformly use `Expression`?

Reply:
Because that alters the circuit 🙂

More specifically, `constant * x` results in two different ASTs depending on which you use. With `Expression` you end up with `AstNode::Mul(Expression, Expression)`, whereas with a bare `pallas::Base` you get `AstNode::Scale(Expression, pallas::Base)`.

The places we've used `Expression` here are where we actually use them in addition, and we only have `impl Add<Expression> for AstNode` defined. We could in theory defined `impl Add<Field> for AstNode` and have it wrap the field element in an `Expression::Constant`, which would move this in the opposite direction to what you suggested, but still be more uniform.